### PR TITLE
New packages python39 and python310

### PR DESCRIPTION
**Pull request by nieder**
Adds new packages python39 and python310.
Python-3.9 is the first upstream to directly support macOS 11 and macOS 12.
We haven't been able to get python38 and earlier to work on  macOS 11/12, so this is the first python that builds and runs on these systems. Latest python-3.8.12 *might* support macOS11/12, but not sure yet.

I removed a lot of the OS-versioned hacks (mostly around finding the SDK and system library cache) that had accumulated into the earlier pythons to work with newer macOS because the underlying issues had been taken care of upstream.

Also cleaned up the lists of tests to just those that currently are known to fail (on macOS12).

Tested on just macOS 12.1.
Please test on anything before 10.14.5.

### CODE REVIEW

**Comment by dmacks** [comment]
On 10.13...

Python39 [coming shortly...github doesn't have a 'save as draft']

Python310: did not detect support for:
* ossaudiodev
* spwd
which is the correct number of missing, and built even though detected, as configured:
* _abc
* pwd
* time
The first set of self-tests report SUCCESS, but there were some "failed (env changed)" messages along the way. The second set of self-tests had hard failure:
```
0:08:10 load avg: 7.30 [383/420/1] test_pkgutil failed (1 error)
test test_pkgutil failed -- Traceback (most recent call last):
  File "/sw/build.build/python310-3.10.1-1/Python-3.10.1/Lib/test/test_pkgutil.py", line 263, in test_name_resolution
    mod = importlib.import_module(uw)
  File "/sw/build.build/python310-3.10.1-1/Python-3.10.1/Lib/importlib/__init__.py", line 126, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1050, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1027, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1004, in _find_and_load_unlocked
ModuleNotFoundError: No module named 'é'
```

**Comment by nieder** [comment]
test_pkgutil test failure : https://bugs.python.org/issue41154
It seems to be an HFS+ failure which is why I didn't see it on APFS. We could just add it to the exclusion list for 3.9 and 3.10 since upstream has deprecated the relevant module.

**Comment by nieder** [comment]
This is my module detection list on 12.1, so your results match:
```
The necessary bits to build these optional modules were not found:
ossaudiodev           spwd                                     
To find the necessary bits, look in setup.py in detect_modules() for the module's name.
(Fink package build should have 2 missing)


The following modules found by detect_modules() in setup.py, have been
built by the Makefile instead, as configured by the Setup files:
_abc                  pwd                   time               
```
And I too see the `failed (env changed)` diagnostics, but I don't think they matter

**Comment by nieder** [comment]
FWIW, python is moving to annual release cycles: https://www.python.org/dev/peps/pep-0602/
Each new release will still be security supported for 5 years.
And python36 reached EOL yesterday.

**Comment by dmacks** [comment]
> test_pkgutil test failure : https://bugs.python.org/issue41154 It seems to be an HFS+ failure which is why I didn't see it on APFS. We could just add it to the exclusion list for 3.9 and 3.10 since upstream has deprecated the relevant module.

Disabling that test gets both python39 and python310 fully built. There is a small .deb difference if I build with vs without tests: with tests, I get a bunch of __pycache__ subdirs in various pythonXX/Tools subdirs. These subdirs are empty, but they do not exist at all if tests=off.

**Comment by nieder** [comment]
Python38 has the same thing with empty `Tools/.../__pycache__` folders. Can use something like `find %i -name *__pycache__* -empty | xargs rmdir` (I've never been good with the -exec or -delete options to find).

File: 10.9-libcxx/stable/main/finkinfo/languages/python310.info [hunk 1]
```diff
@@ -173,6 +173,10 @@ InstallScript: <<
 	# install some docs and other useful tidbits
 	rm -rf Misc/RPM
 	/bin/cp -R Misc Tools %i/lib/python%type_raw[python]
+
+	# remove leftovers from testing
+	# these dirs are empty and only created if tests are run
+	find %i/lib/python%type_raw[python]/Tools -name *__pycache__* -empty | xargs rmdir
```
Comment (dmacks) [inline 10.9-libcxx/stable/main/finkinfo/languages/python310.info:180]: Safer to:
```
find %i/lib/python%type_raw[python]/Tools -name *__pycache__* -empty -type d -delete
```

**Comment by dmacks** [comment]
Safer to
```
find %i/lib/python%type_raw[python]/Tools -name *__pycache__* -empty -type d -delete
```

**Comment by nieder** [comment]
Ugh. For some reason, running `/usr/bin/find ./Tools -name __pycache__ -empty -type d -delete` manually in %b works, but fails in TestScript. Even just `find ...` w/out the -delete doesn't find anything. Definitely in the right directory, etc.

**Comment by nieder** [comment]
The `-empty` flag makes it fail when run in TestScript. And `find ./Tools -name __pycache__ -type d -delete` fails. Makes me think they might not be empty (as far as find is concerned).

**Comment by sth0** [comment]
hidden dot files such as .DS_Store?  Usually find works just fine for me.  It really seems that it is not empty as find seems to not trigger.  Also if there is a symbolic link that has the destination missing it will not show up with some ls flags.  (I think the -L)

**Comment by TheSin-** [comment]
instead of -delete use -exec rm -rf {} \;

**Comment by sth0** [comment]
actually use -delete -exec rm -rf {} \;
The backslash is important and there must be the space after the closing curly bracket.
(and take out the spurious "use" as a typo.

**Comment by nieder** [comment]
Directories are very much empty according to `ls` (and various flags to find hidden files).
```
+ /usr/bin/find ./Tools -name __pycache__ -type d -exec rmdir '{}' ';'
rmdir: ./Tools/c-analyzer/cpython/__pycache__: Directory not empty
rmdir: ./Tools/c-analyzer/c_analyzer/__pycache__: Directory not empty
rmdir: ./Tools/c-analyzer/c_analyzer/parser/__pycache__: Directory not empty
rmdir: ./Tools/c-analyzer/c_analyzer/common/__pycache__: Directory not empty
rmdir: ./Tools/c-analyzer/c_analyzer/variables/__pycache__: Directory not empty
```
This eventually worked:
```
/usr/bin/find %b/Tools -name __pycache__ -type d -depth -exec rm -rf "{}" \;
```
`-depth` needed to keep find from trying to go into the directory after deleting it.

**Comment by dmacks** [comment]
LGTM. I took a look at users of pythonXX and none seem to sort by "major.minor", so there is no priority problem of 3.9 being higher than 3.10. Older pythonXX could benefit from the improved find...rm... identified here, but that's not a reason to wait here.

**Comment by nieder** [comment]
OK. We will also want to restrict py34-py36 to 10.9-10.15.

**Comment by nieder** [comment]
Validation failures with python3.10 (3.9 is OK). Happens on multiple py310 packages.
```
Validating .deb dir /sw/build.build/root-termcolor-py310-1.1.0-1...
Error: File in a language-versioned package does not have a pathname specific to that version.
	Offending file: /sw/lib/python3.10/site-packages/termcolor.py
	Offending file: /sw/lib/python3.10/site-packages/termcolor-1.1.0-py3.10.egg-info
```

**Comment by dhomeier** [comment]
> ```perl
> 			# numbers are all "small" (one-digit)
> 			$langver_re =~ /^(\d)(\d)(\d)$/;
> 			# -pyXY is pythonX.Y
> ```

Yes, I needed something like
```perl
			# numbers are all "small" (one-digit) up to Python 3.10
			$langver_re =~ /^(\d)(1??)(\d)$/;
			# -pyXY is pythonX.Y, -pyX1Y is pythonX.1Y
```
(if we want to support up to Python 3.19 for now ;-)
Otherwise builds and works on 10.14.6 and 10.13.6, thanks!

**Comment by dmacks** [comment]
> Validation failures with python3.10 (3.9 is OK). Happens on multiple py310 packages.
> 
> ```
> Validating .deb dir /sw/build.build/root-termcolor-py310-1.1.0-1...
> Error: File in a language-versioned package does not have a pathname specific to that version.
> 	Offending file: /sw/lib/python3.10/site-packages/termcolor.py
> 	Offending file: /sw/lib/python3.10/site-packages/termcolor-1.1.0-py3.10.egg-info
> ```

Fixed in https://github.com/fink/fink/commit/f7ca63721d485c49d36043e399e99391baf7b121

**Comment by dmacks** [comment]
Heh...my commit landed at the same time @dhomeier was commenting. In terms of code, this is just a string-parsing regexp like scan() to decipher major vs minor, not any sort of validation of the value itself. So it seemed clearer to generalize it as "one digit followed by one or two digits" rather than stricter "...and if three digits, middle must be '1'". If we ever get as high as python10.0, we can figure something else out:)

**Comment by dhomeier** [comment]
Just a suggestion in case we wanted to take the "small numbers" still half-seriously – otherwise would have gone with `(\d??)` as well, but `(\d{1,2})` is rather more elegant.
I have not heard even from a tentative timeline for Python 4.0, so this should last for a while. ;-)

**Comment by nieder** [comment]
Putting it here since it's probably a similar code failure as the validation check:
This is in TestDepends:
`(%type_pkg[python] >= 39)` doesn't evaluate to true with foo-py310. But `(%type_pkg[python] >= 310)` works as expected. Confused because 310 > 39, so I would expect the first to work.

**Comment by dmacks** [comment]
Those conditionals are string-compare not number-compare.

**Pull request merged**